titus-server-master/src/main/java/com/netflix/titus/master/kubernetes/pod/legacy/TitusProvidedContainerEnvFactory.java

```diff
@@ -16,32 +16,45 @@
 
 package com.netflix.titus.master.kubernetes.pod.legacy;
 
+
+import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 
 import com.netflix.titus.api.jobmanager.model.job.BatchJobTask;
 import com.netflix.titus.api.jobmanager.model.job.Job;
 import com.netflix.titus.api.jobmanager.model.job.Task;
+import com.netflix.titus.common.util.tuple.Pair;
 import com.netflix.titus.runtime.kubernetes.KubeConstants;
 
 public class TitusProvidedContainerEnvFactory implements ContainerEnvFactory {
 
     private static final ContainerEnvFactory INSTANCE = new TitusProvidedContainerEnvFactory();
 
     @Override
-    public Map<String, String> buildContainerEnv(Job<?> job, Task task) {
+    public Pair<List<String>, Map<String, String>> buildContainerEnv(Job<?> job, Task task) {
+        // In TitusProvidedContainerEnvFactory, systemEnvNames ends up having everything we add
+        List<String> systemEnvNames = new ArrayList<>();
         Map<String, String> env = new HashMap<>();
         env.put(KubeConstants.POD_ENV_TITUS_JOB_ID, task.getJobId());
+        systemEnvNames.add(KubeConstants.POD_ENV_TITUS_JOB_ID);
```

Comment:
Move these to immutable `List` constant, which is copied only for batch jobs to add the extra entry.